### PR TITLE
Simplify Product Preferences Form

### PR DESCRIPTION
**Pull request by Progi1984**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Simplify Product Preferences Form
| Type?         | refacto
| Category?     | BO
| BC breaks?    | yes
| Deprecations? | no
| Fixed ticket? | Part of #16482
| How to test?  | Simplifying Configuration -> Shop Parameters -> Product Preferences. Everything must look the same aside for help now appearing under inputs instead as blue box, also some fields now will appear as required because they always were.

:notebook: **BC Break**
Backwards compatibility break introduced due to extension of TranslationAwareType by PageType and PaginationType. This means if any module extends those types they will get an exception upon upgrading to PS version containing changes in this PR.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27892)

### CODE REVIEW

**Review by saulaski — DISMISSED** [review]

**Review by matthieu-rolland — DISMISSED** [review]

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralType.php [hunk 1]
```diff
@@ -46,25 +46,78 @@ class GeneralType extends TranslatorAwareType
...
         $builder
-            ->add('catalog_mode', SwitchType::class)
-            ->add('catalog_mode_with_prices', SwitchType::class)
+            ->add('catalog_mode', SwitchType::class, [
+                'label' => $this->trans('Catalog mode', 'Admin.Shopparameters.Feature'),
+                'help' => $this->trans(
+                        'Catalog mode disables the shopping cart on your store. Visitors will be able to browse your products catalog, but not buy them.',
+                        'Admin.Shopparameters.Help'
+                    ) . '<br />' . $this->trans(
+                        'Have specific needs? Edit particular groups to let them see prices or not.',
```
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralType.php]: This help text could be clearer on where to perform the action.

```suggestion
                        'To hide prices for a specific group, go to [1]Customer Settings > Groups[\1].',                        
```

Also, this help text should be placed under the "Show prices" field instead of "Catalog mode".

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralType.php [hunk 2]
```diff
@@ -46,25 +46,78 @@ class GeneralType extends TranslatorAwareType
...
+                ),
                 'required' => false,
             ])
             ->add('short_description_limit', TextWithUnitType::class, [
+                'label' => $this->trans(
+                    'Max size of product summary',
+                    'Admin.Shopparameters.Feature'
+                ),
+                'help' => $this->trans(
+                    'Set the maximum size of the summary of your product description (in characters).',
```
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralType.php]: I think we can remove this tooltip. The label is already clear enough and it's repetitive. + It's not clear: are we talking about the product summary or the description?

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralType.php [hunk 3]
```diff
@@ -46,25 +46,78 @@ class GeneralType extends TranslatorAwareType
...
+                    'Admin.Shopparameters.Feature'
+                ),
+                'help' => $this->trans(
+                    'When active, friendly URL will be updated on every save.',
+                    'Admin.Shopparameters.Help'
+                ),
+            ])
+            ->add('default_status', SwitchType::class, [
+                'label' => $this->trans(
+                    'Default activation status',
```
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralType.php]:
```suggestion
                    'Activate new products by default',
```

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralType.php [hunk 4]
```diff
@@ -46,25 +46,78 @@ class GeneralType extends TranslatorAwareType
...
+                    'Admin.Shopparameters.Feature'
+                ),
+                'help' => $this->trans(
+                    'When active, new products will be activated by default during creation.',
```
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralType.php]: This help text can be removed.

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php [hunk 5]
```diff
@@ -37,27 +37,84 @@
...
-            ->add('allow_add_variant_to_cart_from_listing', SwitchType::class)
+            ->add('display_quantities', SwitchType::class, [
+                'label' => $this->trans(
+                    'Display available quantities on the product page',
+                    'Admin.Shopparameters.Feature'
+                ),
+            ])
+            ->add('display_last_quantities', IntegerType::class, [
+                'label' => $this->trans(
+                    'Display remaining quantities when the quantity is lower than',
```
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php]: I tested that feature but couldn't make it work. Where are the remaining quantities supposed to appear? 🤔

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php [hunk 6]
```diff
@@ -37,27 +37,84 @@
...
                 'choice_translation_domain' => 'Admin.Global',
             ])
-            ->add('display_discount_price', SwitchType::class);
+            ->add('display_discount_price', SwitchType::class, [
+                'label' => $this->trans(
+                    'Display discounted price',
+                    'Admin.Shopparameters.Feature'
+                ),
+                'help' => $this->trans(
+                    'In the volume discounts board, display the new price with the applied discount instead of showing the discount (ie. "-5%").',
```
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php]: What "discounted prices" are we talking about? Specific prices? Because I've just tested the feature and couldn't see the difference between when it's enabled and when it's disabled. 🤔 I don't understand what is the "volume discounts board".

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PaginationType.php [hunk 7]
```diff
@@ -36,16 +36,33 @@
...
     {
         $builder
-            ->add('products_per_page', IntegerType::class)
+            ->add('products_per_page', IntegerType::class, [
+                'label' => $this->trans(
+                    'Products per page',
+                    'Admin.Shopparameters.Feature'
+                ),
+                'help' => $this->trans(
+                    'Number of products displayed per page. Default is 10.',
```
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PaginationType.php]: This help text can be removed, the label speaks for itself.

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PaginationType.php [hunk 8]
```diff
@@ -59,6 +76,14 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'required' => true,
             ])
             ->add('default_order_way', ChoiceType::class, [
+                'label' => $this->trans(
+                    'Default order method',
+                    'Admin.Shopparameters.Feature'
+                ),
+                'help' => $this->trans(
+                    'Default order method for product list.',
```
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PaginationType.php]: This help text can be removed, the label speaks for itself.

**Review by Julievrz — COMMENTED** [review]
A few suggestions and questions about wording

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php (hunk 5 above)
Comment (Progi1984) [reply src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php]: It impacts the configuration `PS_LAST_QTIES`. That impacts in front the label `Last items in stock`.

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php (hunk 6 above)
Comment (Progi1984) [reply src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php]: That impacts the configuration `PS_DISPLAY_DISCOUNT_PRICE`. It used with Specific Prices. And in front, if enabled, that shows the price else the discount. I think you can see it in volumes discounts table on the front office.

image

**Review by matthieu-rolland — DISMISSED** [review]

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php (hunk 6 above)
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php]:
```suggestion
                    'In the volume discount table on the product page, display the discounted unit price instead of the unit discount. E.g. If you sell a product for $10 with a discount of $2 from 3 items purchased, the discounted unit price ($8) will be displayed instead of the unit discount ($2).',
```

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php [hunk 9]
```diff
@@ -37,27 +37,84 @@
...
                     '-' => '-',
                     ',' => ',',
                 ],
                 'required' => true,
                 'choice_translation_domain' => 'Admin.Global',
             ])
-            ->add('display_discount_price', SwitchType::class);
+            ->add('display_discount_price', SwitchType::class, [
+                'label' => $this->trans(
+                    'Display discounted price',
```
Comment (Julievrz) [inline src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php]:
```suggestion
                    'Display the discounted unit price',
```

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php (hunk 6 above)
Comment (Progi1984) [reply src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php]: Fixed

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php (hunk 9 above)
Comment (Progi1984) [reply src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php]: Fixed

**Review by matthieu-rolland — DISMISSED** [review]

**Review by matks — DISMISSED** [review]

File: src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml [hunk 10]
```diff
@@ -1839,17 +1855,3 @@ services:
     public: true
     tags:
       - { name: form.type }
-
-  form.type.common.zone_choice:
```
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml]: Why remove those forms? Is it maybe a typo or a mistake after a rebase (since I believe you worked on these types recently in another PR maybe some contents were mixed)?

It doesn't seem related to the PR anyway so it seems strange.

**Review by jolelievre — COMMENTED** [review]
Thanks @Progi1984 

I just want to confirm with you that the removal of `ZoneChoiceType` and `ConfigurableCountryChoiceType` service definitions is on purpose because it doesn't seem related to the rest of this PR as they are not even used by the improved product form types.

**Review by jolelievre — DISMISSED** [review]

**Comment by jolelievre** [comment]
@matks or @matthieu-rolland can you please re-approve this PR?

**Review by PierreRambaud — DISMISSED** [review]

**Comment by Progi1984** [comment]
@PierreRambaud @jolelievre @matks @matthieu-rolland Rebased. Need some reviews.

**Review by jolelievre — DISMISSED** [review]

**Review by PierreRambaud — DISMISSED** [review]

**Review by marwachelly — COMMENTED** [review]
Hello @Progi1984 ,

I tested the PR and I noticed the following issues:

1. With PR some field are marked as required. Is there any spec related?

image


2. Some field are underlined and lead to the page Shop parameters> Customer settings > Groups

https://user-images.githubusercontent.com/89604350/164002783-930f13a2-6147-4018-827b-84bef0dd5c22.mp4




3. Some help text are missing. Not all blue box are replaced.
image

4. The behavior of the filed "show prices" is changed. According to the spec "It should be displayed only when the catalog mode is enabled"
Before:

https://user-images.githubusercontent.com/89604350/164003176-f1068f08-0839-4f93-8ce2-6f015cc8e377.mp4



With PR:


https://user-images.githubusercontent.com/89604350/164002429-b053c0a1-7e22-4e2b-b961-a1b225aa10a0.mp4



5. The order of fields in "products stock" section is changed
Without PR 
image


with PR
image

6. A toggle button without field in the section "products stock" 
image




7. the drop down of the field "Default order of priority for specific prices " is changed
without PR:
image

With PR:
image

8. The help text and the name of the field "Display discounted price" are changed
Without PR
image
With PR
image

Could you please check? 

Thanks,

**Comment by Progi1984** [comment]
@marwachelly 

1. Fixed (I remove all required fields)
2. Fixed (I have bad closed a `<a>` tag)
3. @Julievrz asked to remove some help texts
  * 1 : https://github.com/PrestaShop/PrestaShop/pull/27892#discussion_r827795130
  * 2 : Fixed
  * 3 : https://github.com/PrestaShop/PrestaShop/pull/27892#discussion_r827816261
  * 4 : https://github.com/PrestaShop/PrestaShop/pull/27892#discussion_r827816429
4. Fixed
5. Fixed
6. Fixed
7. @NeOMakinG Is it the new style ? 
Old style : image
New style : image
9. @Julievrz asked to change the help text : https://github.com/PrestaShop/PrestaShop/pull/27892#discussion_r833452867

**Comment by Progi1984** [comment]
@PierreRambaud & @jolelievre Fixed some problems for @marwachelly

**Review by PierreRambaud — DISMISSED** [review]

**Review by matks — DISMISSED** [review]

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/StockType.php [hunk 11]
```diff
@@ -47,59 +47,151 @@ class StockType extends TranslatorAwareType
...
+                    'Admin.Shopparameters.Feature'
+                ),
+                'help' => $this->trans(
+                    'When selling packs of products, how do you want your stock to be calculated?',
+                    'Admin.Shopparameters.Help'
+                ),
+                'choices' => [
+                    'Decrement pack only.' => 0,
+                    'Decrement products in pack only.' => 1,
+                    'Decrement both.' => 2,
```
Comment (atomiix) [inline src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/StockType.php]: I think these have to be translated

Comment (Progi1984) [reply src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/StockType.php]: The translation is done automatically.

Comment (atomiix) [reply src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/StockType.php]: How? 🤔 Using which domain?

Comment (Progi1984) [reply src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/StockType.php]: Fixed. Check the key `choice_translation_domain`.

**Comment by Progi1984** [comment]
@matks & @atomiix Fixed feedbacks from @atomiix

**Review by atomiix — DISMISSED** [review]

**Review by matks — DISMISSED** [review]

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/StockType.php [hunk 12]
```diff
@@ -47,59 +47,152 @@ class StockType extends TranslatorAwareType
...
+                'label' => $this->trans(
+                    'Default pack stock management',
+                    'Admin.Shopparameters.Feature'
+                ),
+                'help' => $this->trans(
+                    'When selling packs of products, how do you want your stock to be calculated?',
+                    'Admin.Shopparameters.Help'
+                ),
+                'choices' => [
+                    'Decrement pack only.' => 0,
```
Comment (jolelievre) [inline src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/StockType.php]: I think you should use const here `PrestaShop\PrestaShop\Core\Domain\Product\Pack\ValueObject\PackStockType::STOCK_TYPE_PACK_ONLY`

**Review by jolelievre — COMMENTED** [review]
Sorry it was already approved, it's not blocking but I have a suggestion

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/StockType.php (hunk 12 above)
Comment (Progi1984) [reply src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/StockType.php]: Fixed

**Comment by Progi1984** [comment]
@matks @jolelievre @atomiix Fixed feedbacks from @jolelievre

**Review by atomiix — DISMISSED** [review]

**Review by sLorenzini — COMMENTED** [review]
hello @Progi1984,

@khouloudbelguith  and I have tested this PR, and we noticed the following issues:

All selected list on the page have a blank first option :

- Quantity discounts based on
- Separator of attribute anchor on the product links 
- Default pack stock management 
- Default order by 
- Default order method 

Without PR:

selected-list-without-pr

With PR:

selected-list

------------------------------

Issue on number of characters on "Max size of short description" field in product page v2

Change the Max size of short description to 20 and save on product settings page.
Go to BO > Catalog > Products V2 > Edit a product
See error > in the summary section, the max size is not updated

max-size-characters

Set the summary to more than 20 characters and save

max-size-exception

------------------------------

Issue  on "Force update of friendly URL" field in product page v2

Set to yes "Force update of friendly URL" switch button on product settings page.
Go to BO > Catalog > Products V2 > Edit a product
Go on "SEO" tab
Update the name of the product
See error > the "Friendly URL" field is not updated 

With product page V1:

https://user-images.githubusercontent.com/2854834/165105843-e001710c-a820-4744-ba35-7f1b9d557738.mp4

With product page V2:

https://user-images.githubusercontent.com/2854834/165105924-55694d5a-3513-4b7b-8df2-6b0d0251d95f.mp4

------------------------------

Issue  on "Default activation status" field in product page v2

Set to yes "Default activation status" switch button on product settings page.
Go to BO > Catalog > Products V2 > add a product
See error at the bottom > the "Offline" switch button is not active

With product page V1:

https://user-images.githubusercontent.com/2854834/165107223-7682c1f7-f89b-41d9-a1ca-ca024ee05490.mp4

With product page V2:

https://user-images.githubusercontent.com/2854834/165107259-5ec3db16-6756-4df3-9847-2dede75b24ad.mp4


------------------------------

There is a space between the "Default order of priority for specific prices" field and its help text:

space-between-field-and-help-text

------------------------------

We have an issue when we set those fields with too long characters:
- Label of in-stock products
- Label of out-of-stock products with allowed backorders
- Label of out-of-stock products with denied backorders
- Delivery time of in-stock products
- Delivery time of out-of-stock products with allowed backorders

image

image (1)

image (2)

image (3)

------------------------------

We have an issue on pagination block

Go to BO > Product settings page > Pagination Block
Set to 0 the "Products per page" field in product settings page
Go to FO > Home category
See error on top of the page

error-pagination

At the bottom of the page, the showing text must be hidden:

showing-product-category-page

Go to BO > Product settings page > Pagination Block
A help text could be added below the "Products per page" field  to inform that we can remove the pagination if it's set to 0

product-per-page-set-0

------------------------------

When we set an invalid number in those fields for example --50

- Number of days for which the product is considered 'new'
- Display remaining quantities when the quantity is lower than
- Products per page

the alert is moving when we try to scroll

https://user-images.githubusercontent.com/2854834/165111416-4f220443-2cc1-47b1-be6e-4cfc398fb27a.mp4

------------------------------

We have an issue when we tested the product settings page in responsive mode on Mac OS **(with and without PR)**:

With responsive mode in chrome debugger:

responsive-chrome-mac

With iphone mode in chrome debugger:

iphone-chrome-mac

With pixel mode in chrome debugger:

pixel-chrome-mac


Tested on browserstack with iphone screen, linux and windows OS and it is good.

------------------------------

We have an issue on selected list in the product settings pages in Firefox on Mac OS **(with and without PR)**:

On some selected list, when we click on it to change the option, we cannot choose the option on the first click:

https://user-images.githubusercontent.com/2854834/165112658-8b47899c-44ea-4c8e-8c2f-1f28a546b979.mp4

Tested on browserstack on Firefox in windows and linus OS and it is good.


Thank you

**Comment by Progi1984** [comment]
@sLorenzini @khouloudbelguith My report.

* All selected list on the page have a blank first option :
  * Fixed
* Issue on number of characters on "Max size of short description" field in product page v2
  * Not in relation with this PR. Could you create an issue ?
* We have an issue when we set those fields with too long characters:
  * Not in relation with this PR. Could you create an issue ?
* We have an issue on pagination block 
  * Not in relation with this PR. Could you create an issue ?
* We have an issue when we tested the product settings page in responsive mode on Mac OS (with and without PR): 
  * Not in relation with this PR. Could you create an issue ?
* We have an issue on selected list in the product settings pages in Firefox on Mac OS (with and without PR): 
  * Not in relation with this PR. Could you create an issue ?

**Review by jolelievre — APPROVED** [review]

**Review by MeKeyCool — COMMENTED** [review]
Sorry I don't feel comfortable enough to speak about specific symfony form refactoring or about behaviors you updated.

**Review by atomiix — APPROVED** [review]

**Comment by sLorenzini** [comment]
> @sLorenzini @khouloudbelguith My report.
> 
> * All selected list on the page have a blank first option :
>   
>   * Fixed
> * Issue on number of characters on "Max size of short description" field in product page v2
>   
>   * Not in relation with this PR. Could you create an issue ?
> * We have an issue when we set those fields with too long characters:
>   
>   * Not in relation with this PR. Could you create an issue ?
> * We have an issue on pagination block
>   
>   * Not in relation with this PR. Could you create an issue ?
> * We have an issue when we tested the product settings page in responsive mode on Mac OS (with and without PR):
>   
>   * Not in relation with this PR. Could you create an issue ?
> * We have an issue on selected list in the product settings pages in Firefox on Mac OS (with and without PR):
>   
>   * Not in relation with this PR. Could you create an issue ?

@Progi1984 thanks for your report.

All issues not in relation with this PR have been created.

Thank you

**Review by khouloudbelguith — APPROVED** [review]
Hi @Progi1984,

It is ok :heavy_check_mark: 

https://user-images.githubusercontent.com/16067358/166726351-d857d210-74ef-4fa8-b040-37ce5d387fcd.mp4

https://user-images.githubusercontent.com/16067358/166726517-3829b4e0-f45c-4675-9658-412df129b88e.mp4


I have only a difference in the wording
**Without PR**
`Default activation status`
image
**With PR**
`Activate new products by default`
image

Ping @Julievrz what do you think?
If it is ok, I validate this PR.

Thank you!

**Comment by atomiix** [comment]
Thanks @Progi1984

**Comment by Julievrz** [comment]
> Hi @Progi1984,
> 
> It is ok ✔️
> 
>  Part.2.mp4 
>  Part.1.mp4 
> I have only a difference in the wording **Without PR** `Default activation status` image **With PR** `Activate new products by default` image
> 
> Ping @Julievrz what do you think? If it is ok, I validate this PR.
> 
> Thank you!

Too late 😅 Tt would have been better to remove the help text indeed. It's redundant, the label speaks for itself.

**Comment by Julievrz** [comment]
We will do a final review of the wording of the product page anyway because it is really difficult to review as it's broken down into several pieces.
